source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+    az_span * pxHostname;
+    az_span * pxDeviceID;
+
+    if( ( xAzureIoTProvisioningClientHandle == NULL ) || ( pucHubHostname == NULL ) ||
+        ( pulHostnameLength == NULL ) || ( pucDeviceID == NULL ) || ( pulDeviceIDLength == NULL ) )
+    {
+        AZLogError( ( "Provisioning hub get failed: Invalid argument" ) );
+        xReturn = eAzureIoTProvisioningInvalidArgument;
+    }
+    else if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState != azureiotprovisioningWF_STATE_COMPLETE )
+    {
+        AZLogError( ( "Provisioning client state is not in complete state" ) );
+        xReturn = eAzureIoTProvisioningFailed;
+    }
+    else if( xAzureIoTProvisioningClientHandle->_internal.ulLastOperationResult )
+    {
+        xReturn = xAzureIoTProvisioningClientHandle->_internal.ulLastOperationResult;
+    }
+    else
+    {
+        pxHostname = &xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.registration_state.assigned_hub_hostname;
+        pxDeviceID = &xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.registration_state.device_id;
+        ulHostnameLength = ( uint32_t ) az_span_size( *pxHostname );
+        ulDeviceIDLength = ( uint32_t ) az_span_size( *pxDeviceID );
+
+        if( ( *pulHostnameLength < ulHostnameLength ) || ( *pulDeviceIDLength < ulDeviceIDLength ) )
+        {
+            AZLogWarn( ( "Memory passed is not enough to store hub info \r\n" ) );
+            xReturn = eAzureIoTProvisioningFailed;
+        }
+        else
+        {
+            memcpy( pucHubHostname, az_span_ptr( *pxHostname ), ulHostnameLength );
+            memcpy( pucDeviceID, az_span_ptr( *pxDeviceID ), ulDeviceIDLength );
+            *pulHostnameLength = ulHostnameLength;
+            *pulDeviceIDLength = ulDeviceIDLength;
+            xReturn = eAzureIoTProvisioningSuccess;
+        }
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTProvisioningClientResult_t AzureIoTProvisioningClient_GetExtendedCode( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle,
+                                                                               uint32_t * pulExtendedErrorCode )
+{
+    AzureIoTProvisioningClientResult_t xReturn;
+
+    if( ( xAzureIoTProvisioningClientHandle == NULL ) ||
+        ( pulExtendedErrorCode == NULL ) )
+    {
+        AZLogError( ( "Provisioning client extended code get failed : Invalid argument" ) );
+        xReturn = eAzureIoTProvisioningInvalidArgument;
+    }
+    else if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState != azureiotprovisioningWF_STATE_COMPLETE )
+    {
+        AZLogError( ( "Provisioning client state is not in complete state" ) );
+        xReturn = eAzureIoTProvisioningFailed;
+    }
+    else
+    {
+        *pulExtendedErrorCode =
+            xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.registration_state.extended_error_code;
```

Comment:
1 - 
One minor concern I have here is that we can get into STATE_COMPLETE where we think we have the extended error code but in reality failed for some other reason.

I initially thought we could just check `( xAzureIoTProvisioningClientHandle->_internal.lastOperationResult )` in this function before returning `eAzureIoTProvisioningSuccess`.  I don't think we can set the lastOperationResult to some sort of failure *AND* get a DPS result code here?

But that leads to 

2 -- Can we even check on `azureiotprovisioningWF_STATE_COMPLETE` here?  It seems naively that we could get an error from hub saying "Call back again in 2 minutes" and we wouldn't be in COMPLETE, but I think the extended error code would still be set?  This could potentially be meaningful about extnded errors about why I'm being told to call back later so maybe this should be plumbed (assuming DPS even would do a call-back-later and extended code at same time??)

Reply:
Regarding 2: We are checking for complete state : line 921.

---
In reply to: [603557149](https://github.com/Azure/azure-iot-middleware-freertos/pull/49#discussion_r603557149) [](ancestors = 603557149)

Reply:
Correct, I see the `azureiotprovisioningWF_STATE_COMPLETE` comparison.  My question is whether that's the correct check?  Maybe we should allow this for both azureiotprovisioningWF_STATE_COMPLETE and for `azureiotprovisioningWF_STATE_WAITING` also for scenario that service said "300021 code, waitfor=2 minutes".  So app may care about the 300021 status and to determine if they want to go through the wait require loop or not.

Reply:
Currently I have added in doc: only call this once registration is done. Once done, this API will return last error code retrun from Service.